synapse/handlers/appservice.py

```diff
@@ -440,6 +464,76 @@ async def _handle_presence(
 
         return events
 
+    async def _handle_to_device(
+        self,
+        service: ApplicationService,
+        new_token: int,
+        users: Collection[Union[str, UserID]],
+    ) -> List[JsonDict]:
+        """
+        Given an application service, determine which events it should receive
+        from those between the last-recorded typing event stream token for this
+        appservice and the given stream token.
+
+        Args:
+            service: The application service to check for which events it should receive.
+            new_token: The latest to-device event stream token.
+            users: The users that should receive new to-device messages.
+
+        Returns:
+            A list of JSON dictionaries containing data derived from the typing events
+                that should be sent to the given application service.
+        """
+        # Get the stream token that this application service has processed up until
+        from_key = await self.store.get_type_stream_id_for_appservice(
+            service, "to_device"
+        )
+
+        # Filter out users that this appservice is not interested in
+        users_appservice_is_interested_in: List[str] = []
+        for user in users:
+            if isinstance(user, UserID):
+                user = user.to_string()
+
+            if service.is_interested_in_user(user):
+                users_appservice_is_interested_in.append(user)
+
+        if not users_appservice_is_interested_in:
+            # Return early if the AS was not interested in any of these users
+            return []
+
+        # Retrieve the to-device messages for each user
+        recipient_user_id_device_id_to_messages = await self.store.get_new_messages(
+            users_appservice_is_interested_in,
+            from_key,
+            new_token,
+        )
+
+        # According to MSC2409, we'll need to add 'to_user_id' and 'to_device_id' fields
+        # to the event JSON so that the application service will know which user/device
+        # combination this messages was intended for.
+        #
+        # So we mangle this dict into a flat list of to-device messages with the relevant
+        # user ID and device ID embedded inside each message dict.
+        message_payload: List[JsonDict] = []
+        for (
+            user_id,
+            device_id,
+        ), messages in recipient_user_id_device_id_to_messages.items():
+            for message_json in messages:
+                # Remove 'message_id' from the to-device message, as it's an internal ID
+                message_json.pop("message_id", None)
+
+                message_payload.append(
+                    {
+                        "to_user_id": user_id,
+                        "to_device_id": device_id,
+                        **message_json,
+                    }
+                )
```

Comment:
this is more of a matter for the MSC, but you can't just smash together to-device messages and EDUs into the same namespace. cf https://github.com/matrix-org/matrix-doc/pull/2409#discussion_r752211109

Reply:
I've got to say, it's sufficiently broken that I'm not terribly happy landing it in synapse mainline as it currently is.

Reply:
given the other stuff you're changing, I think this is trivial. Just take what you've got and stick it somewhere other than `ephemeral`.

Reply:
ftr the MSC has moved these to a `to_device` array at the top level of the transaction.

Reply:
Ah, OK. I was worried about blocking this PR on changes to the MSC. But given that's now gone through and everyone will be using the MSC as the source of truth, it makes sense to me to update the PR.

Reply:
As of [this set of commits](https://github.com/matrix-org/synapse/pull/11215#commits-pushed-6d68b8a), we now place to-device messages in the right place. That being under `de.sorunome.msc2409.to_device` at the top of the AS transaction, instead of `de.sorunome.msc2409.ephemeral`.